storage/replica.go

```diff
@@ -701,56 +853,67 @@ func (r *Replica) addWriteCmd(ctx context.Context, args proto.Request, wg *sync.
 	// writes, send WriteTooOldError; for reads, update the write's
 	// timestamp. When the write returns, the updated timestamp will
 	// inform the final commit timestamp.
-	if usesTimestampCache(args) {
-		r.Lock()
-		rTS, wTS := r.tsCache.GetMax(header.Key, header.EndKey, header.Txn.GetID())
-		r.Unlock()
+	//
+	// Find the maximum timestamp required to satisfy all requests in
+	// the batch and then apply that to all requests.
+	r.Lock()
+	for i := range bArgs.Requests {
+		args := bArgs.Requests[i].GetValue().(proto.Request)
+		header := args.Header()
+		if usesTimestampCache(args) {
+			rTS, wTS := r.tsCache.GetMax(header.Key, header.EndKey, header.Txn.GetID())
 
-		// Always push the timestamp forward if there's been a read which
-		// occurred after our txn timestamp.
-		if !rTS.Less(header.Timestamp) {
-			header.Timestamp = rTS.Next()
-		}
-		// If there's a newer write timestamp...
-		if !wTS.Less(header.Timestamp) {
-			// If we're in a txn, we still go ahead and try the write since
-			// we want to avoid restarting the transaction in the event that
-			// there isn't an intent or the intent can be pushed by us.
-			//
-			// If we're not in a txn, it's trivial to just advance our timestamp.
-			if header.Txn == nil {
-				header.Timestamp = wTS.Next()
+			// Always push the timestamp forward if there's been a read which
+			// occurred after our txn timestamp.
+			if !rTS.Less(bArgs.Timestamp) {
+				bArgs.Timestamp = rTS.Next()
+			}
+			// If there's a newer write timestamp...
+			if !wTS.Less(bArgs.Timestamp) {
+				// If we're in a txn, we still go ahead and try the write since
+				// we want to avoid restarting the transaction in the event that
+				// there isn't an intent or the intent can be pushed by us.
+				//
+				// If we're not in a txn, it's trivial to just advance our timestamp.
+				if header.Txn == nil {
+					bArgs.Timestamp = wTS.Next()
+				}
 			}
 		}
 	}
 
+	// Make sure the batch timestamp is copied consistently to each request.
+	// TODO(tschottdorf): self-overlap is real and I think we should deal with
```

Comment:
reworded.

> 	// Make sure the batch timestamp is copied consistently to each request.
	// TODO(tschottdorf): Forward() is provisional, depending on the outcome
	// of the self-overlap discussion. See setBatchTimestamps.